### PR TITLE
Added auth, userlogs, removed future targets details and many unit tests

### PR DESCRIPTION
**Pull request by fjoubert**
* Added login and logout messages
* Added a way to create and list userlogs and list userlog tags after a login
* Removed future_targets_detail, we are supplying the future targets as per https://github.com/ska-sa/katcorelib/pull/206, so we do not need to supply catalogues
* Added many unit tests, coverage at 93%
* And as usual - pep8 fixes

### CODE REVIEW

File: katportalclient/client.py [hunk 1]
```diff
@@ -1421,6 +1345,249 @@ def sensors_histories(self, filters, start_time_sec, end_time_sec, include_value
...
+        and end_time. For example of an userlog has a start_time before the
+        given start_time and an end time after the given end_time, the time
+        window of that userlog intersects with the time window of the given
+        start_time and end_time.
+
+        If an userlog has no end_time, an end_time of infinity is assumed.
+        For example, if the given end_time is after the userlog's start time,
+        there is an intersection of the two time windows.
+
+        Here are some visual representations of the time window intersections:
```
Comment (martinslabber) [inline katportalclient/client.py]: Cool text graphics.

File: katportalclient/test/test_client.py [hunk 2]
```diff
@@ -779,147 +984,296 @@ def test_future_targets(self):
...
+            'attachment_count': 0,
+            'id': 40,
+            'end_time': '2017-02-07 23:59:59'
+        }
+        userlog = yield self._portal_client.modify_userlog(userlog_to_modify, [1,2,3])
         self.assertEquals(
-            result,
-            [{u'name': u'Moon', u'testvar': u'test variable 1'},
-             {u'name': u'Sun', u'testvar': u'test variable 2'}])
+            userlog,
```
Comment (ajoubertza) [inline katportalclient/test/test_client.py]: This doesn't really test that the expected modification happened, just that there was a call to authorized_fetch, which is returning your own mock content.  E.g. the [1, 2, 3] tag_ids don't show up in the actual output.  Suggest you also check the `body` param passed to that mock call is the modified version of the user log.

File: katportalclient/client.py [hunk 3]
```diff
@@ -407,10 +527,11 @@ def _websocket_message(self, msg):
...
             self._logger.warn("Websocket server disconnected!")
             if not self._disconnect_issued:
-                self._ws.close()
-                self._ws = None
+                if self._ws is not None:
+                    self._ws.close()
+                    self._ws = None
                 yield self._connect(reconnecting=True)
-            return
+            raise tornado.gen.Return()
```
Comment (martinslabber) [inline katportalclient/client.py]: Why did you make this change? It is not needed/is no different. 
And personally I find a return prettier. You dont have to change anything, I am just curious.

From Tornado Src:
> Coroutines may "return" by raising the special exception
>     `Return(value) <Return>`.  In Python 3.3+, it is also possible for
>     the function to simply use the ``return value`` statement (prior to
>     Python 3.3 generators were not allowed to also return values).
>     In all versions of Python a coroutine that simply wishes to exit
>     early may use the ``return`` statement without a value.
and after some more bla bla bla bla
> By analogy with the return statement, the value argument is optional,
> but it is never necessary to ``raise gen.Return()``.  The ``return``
> statement can be used with no arguments instead.

File: katportalclient/client.py [hunk 4]
```diff
@@ -1421,6 +1345,249 @@ def sensors_histories(self, filters, start_time_sec, end_time_sec, include_value
...
+
+        Userlog:            *--------------------------------------*
+        Search params:              [-----------------]
+                                  Start              End
+        Parameters
+        ----------
+        start_time: str
+            A formatted datetime string used as the start of the time window
+            to query. Format: %Y-%m-%d %H:%m:%s.
+            Default: Today (local time) at %Y-%m-%d 00:00:00
```
Comment (martinslabber) [inline katportalclient/client.py]: Definately say that time is in UTC, repeatedly in the doc strings and the examples.
Someone will email and ask "is time UTC or my local timezone"
Hope your using a strong and smart time parser on the server. Not the python built in.

File: examples/example_userlogs_usage.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    print "There are %s userlog tags." % len(tags)
+    print "=============================="
+    print "Here is a list of userlogs for today:"
+    print userlogs
+
+    # To create an userlog use the following code
+    # To add tags, make an array of tag id's
+    userlog_tags_to_add = [tags[0].get('id'), tags[1].get('id')]
+    userlog_content = "This is where you would put the content of the userlog!"
+    # Start time and end times needs to be in this format 'YYYY-MM-DD HH:mm:ss'
```
Comment (fjoubert) [inline examples/example_userlogs_usage.py]: Added in ad99d45

File: examples/example_userlogs_usage.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        tag_ids=userlog_tags_to_add,
+        start_time=start_time,
+        end_time=end_time)
+
+    print "=============================="
+    print "Created a userlog! This is the new userlog: "
+    print userlog_created
+    print "=============================="
+
+    # To edit an existing userlog, user edit_userlog with the modified userlog
```
Comment (fjoubert) [inline examples/example_userlogs_usage.py]: Fixed in ad99d45

File: examples/example_userlogs_usage.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    userlog_to_edit = userlogs[-1] if userlogs else None
```
Comment (fjoubert) [inline examples/example_userlogs_usage.py]: A fair point, changed it to use the newly created userlog in ad99d45

File: katportalclient/client.py [hunk 8]
```diff
@@ -131,6 +172,78 @@ def __init__(self, url, on_update_callback, io_loop=None, logger=None):
         self._ws_jsonrpc_cache = []
         self._heart_beat_timer = PeriodicCallback(
             self._send_heart_beat, WS_HEART_BEAT_INTERVAL)
+        self._current_user_id = None
+
+    @tornado.gen.coroutine
+    def logout(self):
+        """ Logs user out of katportal. Katportal then deletes the cached
+        session_id for this client. In order to call HTTP requests that
+        requires authentication, the user will need to login in again.
```
Comment (fjoubert) [inline katportalclient/client.py]: Fixed in ad99d45

File: katportalclient/client.py [hunk 9]
```diff
@@ -131,6 +172,78 @@ def __init__(self, url, on_update_callback, io_loop=None, logger=None):
...
+        """
+        try:
+            url = self.sitemap['authorization'] + '/user/logout'
+            response = yield self.authorized_fetch(
+                url=url, auth_token=self._session_id, method='POST', body='{}')
```
Comment (fjoubert) [inline katportalclient/client.py]: Good point, if you don't have a local `_session_id`, then there's no point to calling the katportal endpoint. Added a check in ad99d45

File: katportalclient/client.py [hunk 10]
```diff
@@ -131,6 +172,78 @@ def __init__(self, url, on_update_callback, io_loop=None, logger=None):
...
+
+                self._logger.info('Succesfully logged in as %s',
+                                  response_json.get('email'))
+            else:
+                self._session_id = None
+                self._current_user_id = None
+                self._logger.error('Error in logging see response %s',
+                                   response)
+        except Exception:
+            self._logger.exception('Error in response')
```
Comment (fjoubert) [inline katportalclient/client.py]: It would be better to clear them on an exception. Added in

File: katportalclient/client.py [hunk 11]
```diff
@@ -339,7 +458,7 @@ def _cache_jsonrpc_request(self, jsonrpc_request):
                     requests_to_remove.append(req)
         elif (jsonrpc_request.method.startswith('set_sampling_strat') and
               jsonrpc_request.params[2] == 'none'):
-            # index 2 of params is always the sampling strategy
+            # index 1 of params is always the sampling strategy
```
Comment (fjoubert) [inline katportalclient/client.py]: I was mistaken when I started testing this and never reverted the comment. Should be index 2. Changed inad99d45

File: katportalclient/client.py (hunk 4 above)
Comment (fjoubert) [inline katportalclient/client.py]: Updated the comments to be clearer and corrected the input format in ad99d45

File: katportalclient/client.py [hunk 12]
```diff
@@ -1421,6 +1345,249 @@ def sensors_histories(self, filters, start_time_sec, end_time_sec, include_value
...
+        content: str
+            The content of the userlog, could be any text. Required.
+
+        tag_ids: list
+            A list of tag id's to link to this userlog. Optional.
+            Example: [1, 2, 3, ..]
+
+        start_time: str
+            A formatted datetime string used as the start time of the userlog.
+            Format: %Y-%m-%d %H:%m:%s. Optional.
```
Comment (fjoubert) [inline katportalclient/client.py]: Ditto ad99d45

File: katportalclient/client.py [hunk 13]
```diff
@@ -1421,6 +1345,249 @@ def sensors_histories(self, filters, start_time_sec, end_time_sec, include_value
...
+            'user': self._current_user_id,
+            'content': content
+        }
+        if start_time is not None:
+            new_userlog['start_time'] = start_time
+        if end_time is not None:
+            new_userlog['end_time'] = end_time
+        if tag_ids is not None:
+            new_userlog['tag_ids'] = tag_ids
+
```
Comment (fjoubert) [inline katportalclient/client.py]: Not the content, no. The tag_ids are used to select tags from the database by id, so if this is not an array of ints, it fails to update/add.

File: katportalclient/test/test_client.py [hunk 14]
```diff
@@ -779,147 +984,296 @@ def test_future_targets(self):
...
-                sb_id_code_3,
-                sb_id_code_4])
+    def test_create_jwt_login_token(self):
+        """Test that our jwt encoding works as expected"""
+        test_token = create_jwt_login_token(
+            email='test@test.test', password='testpassword')
+        self.assertEquals(
+            test_token,
+            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6InRlc3RAdGVzdC'
+            '50ZXN0In0.aI9/c3tgy5kaKUMfeVHn/3CWLddz4lZI4yFAqHq/JH0=')
```
Comment (fjoubert) [inline katportalclient/test/test_client.py]: Added some comments in ad99d45 for clarification

File: katportalclient/test/test_client.py (hunk 2 above)
Comment (fjoubert) [inline katportalclient/test/test_client.py]: Excellent point. Added a test for that in ad99d45 for the modify and the create

File: katportalclient/client.py (hunk 3 above)
Comment (fjoubert) [inline katportalclient/client.py]: Yip, this was silly. Reverted in ad99d45

**Pull request merged**